sql/files/defaultdata/compare/compare.cc

```diff
@@ -72,33 +73,65 @@ FILE *openfeedback(const char *feedbackdir, const char *feedback, const char *wh
 	return res;
 }
 
-/* Test two numbers for equality, accounting for +/-INF, NaN and
- * precision. Float f2 is considered the reference value for relative
- * error.
- */
-int equal(flt f1, flt f2, flt float_abs_tol, flt float_rel_tol)
+// The behavior of std::tolower on (signed) char is undefined.
+char tolower_char(char c)
+{
+    return static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
+}
+
+bool equal_case_insensitive(std::string a, std::string b)
 {
-	flt absdiff, reldiff;
+	for (char &c : a) c = tolower_char(c);
+	for (char &c : b) c = tolower_char(c);
+
+	return a==b;
+}
+
+/* Test two floating-point numbers for equality, accounting for +/-INF, NaN, and precision.
+ * Float `jval` is considered the reference value for relative error.
+ */
+void compare_float(std::string judge, std::string team, flt jval, flt tval, flt float_abs_tol, flt float_rel_tol, std::string extra_msg) {
 	/* Finite values are compared with some tolerance */
-	if ( std::isfinite(f1) && std::isfinite(f2) ) {
-		absdiff = fabsl(f1-f2);
-		reldiff = fabsl((f1-f2)/f2);
-		return !(absdiff > float_abs_tol && reldiff > float_rel_tol);
-	}
+	if (std::isfinite(tval) && std::isfinite(jval)) {
+		flt absdiff = fabsl(tval-jval);
+		flt reldiff = fabsl((tval-jval)/jval);
+		if (float_abs_tol >= 0 && float_rel_tol >= 0) {
+			if (absdiff > float_abs_tol || reldiff > float_rel_tol) {
```

Comment:
After running this validator on BAPC and talking to @mpsijm over a Slack Huddle at the same time, we found out this should be:

```suggestion
			if (absdiff > float_abs_tol && reldiff > float_rel_tol) {
```

otherwise it fails if a submission is wrong EITHER relative or absolute instead or BOTH.

Reply:
And just for completeness: this makes it consistent with [Kattis' default validator](https://github.com/Kattis/problemtools/blob/develop/support/default_validator/default_validator.cc#L150-L151). Note that they write `!( <= )` instead of simply `>`, but it should indeed be the same now. Thanks for finding this one! :smile: